recursively transformDates

anti-h3r0 opened the pull request:
## Description

adjusted `transformDates` function to check recursively for the `toDate()` funciton inside of any objects within the `serializedDocument`.

I'm assuming here that `toDate()` is a property inside an object.

Closes #1 .

- Link(s) to related 3rd party documentation, solved issue(s), commit(s) & PR(s)

JpCapdevila reviewed (changes requested):
@anti-h3r0 I believe it won't work as expected.

Let's say we have this serializedDocument:

```js
let serializedDocument = {
  data: {
    userData: {
      signIn: firestoreTimestamp
    },
    posts: [
      {
        createdAt: firestoreTimestamp
      }
    ]
  }
}
```

When we call transformDates recursively, the object passed will never have data on it, and because we are iterating over `Object.keys(serializedDocument.data)`, nothing nested will be transformed.

Also, I don't think the code handles nested objects like `userData` in the example.

Please correct me if I'm wrong. In any case maybe it would be good to have some unit tests here.

anti-h3r0 commented:
@JpCapdevila kk, I fixed it and added a helper function. 

## note : 
arrays of objects must have an object property which contains the toDate() function. please let me know if this behavior needs to be altered. I'm not quite sure what format to expect

## for example
``` javascript
let x = {
    posts: [
        {
            createdAt: "originalTime",
            toDate() {
                console.log("success");
                return new Date();
            }
        },
        {
            createdAt: {
                dateTime: "dateTime",
                toDate() {
                    console.log(`calling toDate() on ${this.dateTime}`);
                    return new Date();
                }
            }
        }
    ]
};

transformDatesHelper(x);  // will return
// {
//     "posts": [
//         {
//             "createdAt": "originalTime"
//         },
//         {
//             "createdAt": "2021-05-18T17:06:37.590Z"
//         }
//     ]
// }
```

anti-h3r0 commented:
## Proposed Changes
if I were to add a quick check like so : 

``` javascript 
function transformDatesHelper(data) {
    if (typeof data.toDate === 'function')
        return data.toDate();

    Object.keys(data).forEach(property => { // ...
```
x would return 
```
// {
//     "posts": [
//         "2021-05-18T17:07:09.619Z",
//         {
//             "createdAt": "2021-05-18T17:07:09.619Z"
//         }
//     ]
// }
```

## consequences
no recursive iteration inside any object which contains a toDate(). This is technically already the case though, I guess. Do you have any thoughts @JpCapdevila @i-amcharles ?

anti-h3r0 commented:
## test cases
testing recursive objects, arrays,  sub arrays & objects w/out a toDate() function
am I missing any test cases?

note : ctr+f 'test' and every test in the return should (and does) have a date attached

``` javascript
let testObject = {
    data: {
        primaryObj_test1: {
            toDate () {
                return new Date();
            }
        },
        primaryObj_test2: {
            toDate () {
                return new Date();
            }
        },
        recursiveObj: {
            data: "this has no toDate function",
            recursiveObj_SubObj: {
                data: "this has no toDate function",
                recursiveObj_SubObj_test1: {
                    toDate () {
                        return new Date();
                    }
                },
                recursiveObj_SubObj_test2: {
                    toDate () {
                        return new Date();
                    }
                },
                recursiveObj_SubObj_test3: {
                    toDate () {
                        return new Date();
                    }
                },
                recursiveObj_SubObj_SubObj: {
                    recursiveObj_subObj_subObj_test1: {
                        toDate () {
                            return new Date();
                        }
                    },
                    recursiveObj_subObj_subObj_test2: {
                        toDate () {
                            return new Date();
                        }
                    }
                }
            }
        },
        array: [
            {
                array_test1: {
                    toDate () {
                        return new Date();
                    }
                }
            },
            {
                array_test2: {
                    toDate () {
                        return new Date();
                    }
                },
                arrayObj2_SubArray: [
                    {
                        arrayObj2_SubArray_test1: {
                            toDate() {
                                return new Date();
                            }
                        },
                        arrayObj2_SubArray_SubObj: {
                            arrayObj2_SubArray_SubObj_test1: {
                                toDate() {
                                    return new Date();
                                }
                            },
                            arrayObj2_SubArray_SubObj_test2: {
                                toDate() {
                                    return new Date();
                                }
                            }
                        }
                    }
                ]
            },
            [
                {
                    array_SubArray_test1: {
                        toDate() {
                            return new Date();
                        }
                    },
                    array_SubArray_SubObj: {
                        array_SubArray_SubObj_test1: {
                            toDate() {
                                return new Date();
                            }
                        },
                        array_SubArray_SubObj_SubObj: {
                            array_SubArray_SubObj_SubObj_test1: {
                                toDate() {
                                    return new Date();
                                }
                            },
                        }
                    }
                },
                {
                    array_SubArray_test2: {
                        toDate() {
                            return new Date();
                        }
                    }
                }
            ]
        ]
    }
}

console.log(JSON.stringify(transformDates(testObject), null, 2));

// {
//     "primaryObj_test1": "2021-05-19T15:56:46.640Z",
//     "primaryObj_test2": "2021-05-19T15:56:46.640Z",
//     "recursiveObj": {
//         "data": "this has no toDate function",
//         "recursiveObj_SubObj": {
//             "data": "this has no toDate function",
//             "recursiveObj_SubObj_test1": "2021-05-19T15:56:46.640Z",
//             "recursiveObj_SubObj_test2": "2021-05-19T15:56:46.640Z",
//             "recursiveObj_SubObj_test3": "2021-05-19T15:56:46.640Z",
//             "recursiveObj_SubObj_SubObj": {
//                 "recursiveObj_subObj_subObj_test1": "2021-05-19T15:56:46.640Z",
//                 "recursiveObj_subObj_subObj_test2": "2021-05-19T15:56:46.640Z"
//             }
//         }
//     },
//     "array": [
//         {
//             "array_test1": "2021-05-19T15:56:46.640Z"
//         },
//         {
//             "array_test2": "2021-05-19T15:56:46.640Z",
//             "arrayObj2_SubArray": [
//                 {
//                     "arrayObj2_SubArray_test1": "2021-05-19T15:56:46.640Z",
//                     "arrayObj2_SubArray_SubObj": {
//                         "arrayObj2_SubArray_SubObj_test1": "2021-05-19T15:56:46.640Z",
//                         "arrayObj2_SubArray_SubObj_test2": "2021-05-19T15:56:46.640Z"
//                     }
//                 }
//             ]
//         },
//         [
//             {
//                 "array_SubArray_test1": "2021-05-19T15:56:46.640Z",
//                 "array_SubArray_SubObj": {
//                     "array_SubArray_SubObj_test1": "2021-05-19T15:56:46.640Z",
//                     "array_SubArray_SubObj_SubObj": {
//                         "array_SubArray_SubObj_SubObj_test1": "2021-05-19T15:56:46.640Z"
//                     }
//                 }
//             },
//             {
//                 "array_SubArray_test2": "2021-05-19T15:56:46.640Z"
//             }
//         ]
//     ]
// }
```

JpCapdevila reviewed (approved):
LGTM!
